### ISSUE TITLE
rkyv 0.4.0 doesn't work on no_std

### ISSUE THREAD

**Issue by xobs** [djkoloski/rkyv#69]
When building with `no_std`, rkyv `0.4.0` fails to compile:

```
   Compiling rkyv v0.4.0
error[E0433]: failed to resolve: use of undeclared type `S`
   --> C:\Users\smcro\.cargo\registry\src\github.com-1ecc6299db9ec823\rkyv-0.4.0\src\core_impl\mod.rs:429:70
    |
429 |     fn serialize_unsized(&self, serializer: &mut S) -> Result<usize, S::Error> {
    |                                                                      ^ use of undeclared type `S`

error[E0433]: failed to resolve: use of undeclared crate or module `slice`
   --> C:\Users\smcro\.cargo\registry\src\github.com-1ecc6299db9ec823\rkyv-0.4.0\src\core_impl\mod.rs:431:25
    |
431 |             let bytes = slice::from_raw_parts(
    |                         ^^^^^ use of undeclared crate or module `slice`

error[E0433]: failed to resolve: use of undeclared crate or module `mem`
   --> C:\Users\smcro\.cargo\registry\src\github.com-1ecc6299db9ec823\rkyv-0.4.0\src\core_impl\mod.rs:433:30
    |
433 |                 self.len() * mem::size_of::<T>(),
    |                              ^^^ use of undeclared crate or module `mem`

error[E0433]: failed to resolve: use of undeclared type `S`
   --> C:\Users\smcro\.cargo\registry\src\github.com-1ecc6299db9ec823\rkyv-0.4.0\src\core_impl\mod.rs:443:88
    |
443 |     fn serialize_metadata(&self, serializer: &mut S) -> Result<Self::MetadataResolver, S::Error> {
    |                                                                                        ^ use of undeclared type `S`

error[E0412]: cannot find type `S` in this scope
   --> C:\Users\smcro\.cargo\registry\src\github.com-1ecc6299db9ec823\rkyv-0.4.0\src\core_impl\mod.rs:428:63
    |
428 | impl<T: ArchiveCopy, D: Serializer + ?Sized> SerializeUnsized<S> for [T] {
    |                      -                                        ^ help: a type parameter with a similar name exists: `D`
    |                      |
    |                      similarly named type parameter `D` defined here

error[E0412]: cannot find type `S` in this scope
   --> C:\Users\smcro\.cargo\registry\src\github.com-1ecc6299db9ec823\rkyv-0.4.0\src\core_impl\mod.rs:429:50
    |
428 | impl<T: ArchiveCopy, D: Serializer + ?Sized> SerializeUnsized<S> for [T] {
    |                      - similarly named type parameter `D` defined here
429 |     fn serialize_unsized(&self, serializer: &mut S) -> Result<usize, S::Error> {
    |                                                  ^ help: a type parameter with a similar name exists: `D`

error[E0412]: cannot find type `S` in this scope
   --> C:\Users\smcro\.cargo\registry\src\github.com-1ecc6299db9ec823\rkyv-0.4.0\src\core_impl\mod.rs:443:51
    |
428 | impl<T: ArchiveCopy, D: Serializer + ?Sized> SerializeUnsized<S> for [T] {
    |                      - similarly named type parameter `D` defined here
...
443 |     fn serialize_metadata(&self, serializer: &mut S) -> Result<Self::MetadataResolver, S::Error> {
    |                                                   ^ help: a type parameter with a similar name exists: `D`

error[E0207]: the type parameter `D` is not constrained by the impl trait, self type, or predicates
   --> C:\Users\smcro\.cargo\registry\src\github.com-1ecc6299db9ec823\rkyv-0.4.0\src\core_impl\mod.rs:428:22
    |
428 | impl<T: ArchiveCopy, D: Serializer + ?Sized> SerializeUnsized<S> for [T] {
    |                      ^ unconstrained type parameter

error: aborting due to 8 previous errors

Some errors have detailed explanations: E0207, E0412, E0433.
For more information about an error, try `rustc --explain E0207`.
error: could not compile `rkyv`

To learn more, run the command again with --verbose.
```

**Comment by djkoloski** [comment]
Thanks, I'll look into this as soon as possible.

**Comment by djkoloski** [comment]
I thought this was being tested by the `rkyv_test` crate, but it turns out that I forgot to mark `rkyv_dyn` and `rkyv_typename` as optional dependencies, and even when the test crate didn't request the `std` feature, one of those did and caused the test crate to compile `rkyv` with `std` anyway.

Fixed by a2fcaa60ae7cffa191c02a826b51133f6128ae87, released in 0.4.1.

**Comment by xobs** [comment]
Awesome, thanks for that!

I've been bitten by quirks in the resolver before, and `resolver="2"` seems to solve a lot of these surprising edge cases. It will be stabilized in 1.51 in two weeks, and will become the default in 2021.
